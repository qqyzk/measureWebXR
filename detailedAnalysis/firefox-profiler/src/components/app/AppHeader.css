/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.appHeader {
  display: flex;
  justify-content: space-between;
  padding-bottom: 0.5em;
  border-bottom: 1px solid #ddd;

  /* Negative margin is here to compensate for the children's margins */
  margin: 0 -0.2em 0.5em;
  font-size: 1.5em;
}

.appHeaderLink {
  color: inherit;
  text-decoration: none;
}

/* Note that :focus-visible styles are defined in focus.css */
.appHeaderLink:hover {
  text-decoration: underline;
}

.appHeaderSlogan {
  margin: 0 0.2em;
}

.appHeaderSubtext {
  font-weight: normal;
}

.appHeaderGithubIcon {
  margin: 0 0.2em;
  transition: opacity 250ms var(--animation-curve);
}

.appHeaderGithubIcon svg {
  vertical-align: middle;
}

.appHeaderGithubIcon:hover {
  opacity: 0.5;
}
