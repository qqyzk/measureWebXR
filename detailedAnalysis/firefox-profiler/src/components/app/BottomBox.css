/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.bottom-box {
  display: flex;
  height: 100%;
  flex-flow: column;
}

.bottom-main {
  position: relative;
  display: flex;
  min-height: 0;
  flex: 1;
  flex-flow: column;
  background: var(--grey-20);
}

.bottom-box-bar {
  display: flex;
  overflow: hidden;
  height: 24px;
  flex-flow: row;
  align-items: center;
  justify-content: space-between;
  background: var(--grey-10);
  box-shadow: inset 0 -1px var(--grey-30);
  line-height: 18px;
}

.bottom-box-title {
  margin: 0;
  margin-left: 8px;
  font: inherit;
}

.bottom-close-button {
  width: 24px;
  height: 24px;
  flex-shrink: 0;
  background: url(/res/img/svg/close-dark.svg) no-repeat center / 16px 16px;
}

.sourceStatusOverlay {
  /**
   * Put the overlay on top of everything in .bottom-main, but centered
   * horizontally and vertically. We center using margin: auto, and enforce
   * maximum sizes so that the line numbers and timings are still visible.
   */

  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  display: flex;
  width: max-content;
  min-width: 200px;
  max-width: min(500px, 100%);
  height: min-content;
  flex-flow: row;
  align-items: center;
  padding: 15px;
  border-radius: 10px;
  margin: auto;
  background-color: rgb(240 240 240 / 0.8);
  gap: 15px;
  word-break: break-word;
}

/* Use :before to add a loading spinner image */
.sourceStatusOverlay.loading::before {
  display: block;
  width: 32px;
  height: 32px;
  flex-shrink: 0;
  background: url(/res/img/spinner.png) center center / 32px 32px no-repeat;
  content: '';
}

/* Use :before to add an alert icon */
.sourceStatusOverlay.error::before {
  display: block;
  width: 50px;
  height: 50px;
  flex-shrink: 0;
  background: url(/res/img/svg/error.svg) center center / 32px 32px no-repeat;
  content: '';
  filter: brightness(70%) drop-shadow(0 1px rgb(255 255 255 / 0.5));
}

.sourceStatusOverlay ul {
  padding-left: 20px;
}
