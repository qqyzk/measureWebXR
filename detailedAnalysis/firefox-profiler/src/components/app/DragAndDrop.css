/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.dragAndDropArea {
  display: inherit;
  width: 100%;
  isolation: isolate;
}

.dragAndDropOverlayWrapper {
  position: absolute;
  top: 0;
  left: 0;
  display: flex;
  width: 100%;
  height: 100%;
  box-sizing: border-box;
  align-items: stretch;
  justify-content: center;
  padding: 1em;
  background-color: rgb(255 255 255 / 0.9);
  opacity: 0;
  pointer-events: none;
  transition: opacity 200ms;
}

.dragAndDropOverlayWrapper.dragging {
  opacity: 1;
}

.dragAndDropOverlay {
  display: flex;
  width: 100%;
  align-items: center;
  justify-content: center;
  border: 2px dashed #888;
  border-radius: 10px;
  color: #666;
  font-size: 30px;
  text-align: center;
}
