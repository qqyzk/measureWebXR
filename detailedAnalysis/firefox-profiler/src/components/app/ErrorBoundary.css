.appErrorBoundary {
  display: flex;
  overflow: hidden;
  box-sizing: border-box;
  flex: 1;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  padding: 10px;
  background-color: var(--grey-20);
}

.appErrorBoundaryContents {
  display: flex;
  width: 100%;
  max-width: 600px;
  max-height: calc(100%);
  box-sizing: border-box;
  flex-direction: column;
}

.appErrorBoundaryMessage {
  flex: none;
}

.appErrorBoundaryDetails {
  padding: 25px;
  border-radius: 4px;
  margin: 12px 0;
  background: #fff;
  box-shadow: 0 0 10px rgb(0 0 0 / 0.1);
  color: var(--grey-70);
  font-family: monospace;
  line-height: 1.3;
  opacity: 1;
  overflow-y: auto;

  /* Transition opacity, but do not delay changing the visibility. */
  transition: opacity 150ms, visibility 0s;
  visibility: visible;
  white-space: pre-wrap;
}

.appErrorBoundaryDetails.hide {
  opacity: 0;

  /* Transition opacity, and create a delay on hiding the visibility to allow
     time for the fade out. */
  transition: opacity 150ms, visibility 0s 150ms;
  visibility: hidden;
}
