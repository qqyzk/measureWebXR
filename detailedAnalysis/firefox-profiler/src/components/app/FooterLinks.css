.appFooterLinks {
  /* Position */
  position: fixed;
  right: 0;
  bottom: 0;

  /* Box */
  padding: 0 3px;

  /* Other */
  border-top: 1px solid var(--grey-40);
  border-left: 1px solid var(--grey-40);
  background-color: var(--grey-20);
}

.appFooterLinksLink,
.appFooterLinksClose {
  display: inline-block;
  margin: 2px 4px;
  color: var(--blue-60);
  text-decoration: none;
}

.appFooterLinksClose {
  border: none;
  margin: 0;
  background: none;
  cursor: pointer;
  font: inherit;
}

.appFooterLinksLink:hover,
.appFooterLinksClose:hover {
  text-decoration: underline;
}

.appFooterLinksLanguageSwitcher {
  margin: 2px 5px;
  color: inherit;
  font: inherit;
}

@media (max-width: 767px) {
  .appFooterLinks {
    width: 100%;
    text-align: center;
  }
}
