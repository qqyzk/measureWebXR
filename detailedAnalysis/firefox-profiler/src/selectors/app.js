/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { createSelector } from 'reselect';

import {
  getDataSource,
  getSelectedTab,
  getTimelineTrackOrganization,
  getHiddenGlobalTracks,
  getHiddenLocalTracksByPid,
  getIsActiveTabResourcesPanelOpen,
} from './url-state';
import {
  getGlobalTracks,
  getLocalTracksByPid,
  getActiveTabTimeline,
} from './profile';
import { getZipFileState } from './zipped-profiles.js';
import { assertExhaustiveCheck, ensureExists } from '../utils/flow';
import {
  FULL_TRACK_SCREENSHOT_HEIGHT,
  ACTIVE_TAB_TRACK_SCREENSHOT_HEIGHT,
  TRACK_NETWORK_HEIGHT,
  TRACK_MEMORY_HEIGHT,
  TRACK_IPC_HEIGHT,
  TRACK_PROCESS_BLANK_HEIGHT,
  TIMELINE_RULER_HEIGHT,
  TRACK_VISUAL_PROGRESS_HEIGHT,
  ACTIVE_TAB_TIMELINE_RESOURCES_HEADER_HEIGHT,
  TRACK_EVENT_DELAY_HEIGHT,
  TIMELINE_MARGIN_LEFT,
  ACTIVE_TAB_TIMELINE_MARGIN_LEFT,
  TRACK_PROCESS_CPU_HEIGHT,
} from '../app-logic/constants';

import type {
  AppState,
  AppViewState,
  UrlSetupPhase,
  Selector,
  CssPixels,
  ThreadsKey,
  ExperimentalFlags,
  UploadedProfileInformation,
} from 'firefox-profiler/types';
import type { TabSlug } from 'firefox-profiler/app-logic/tabs-handling';
import type {
  BrowserConnectionStatus,
  BrowserConnection,
} from 'firefox-profiler/app-logic/browser-connection';

/**
 * Simple selectors into the app state.
 */
export const getApp: Selector<AppState> = (state) => state.app;
export const getView: Selector<AppViewState> = (state) => getApp(state).view;
export const getUrlSetupPhase: Selector<UrlSetupPhase> = (state) =>
  getApp(state).urlSetupPhase;
export const getHasZoomedViaMousewheel: Selector<boolean> = (state) => {
  return getApp(state).hasZoomedViaMousewheel;
};
export const getIsSidebarOpen: Selector<boolean> = (state) =>
  getApp(state).isSidebarOpenPerPanel[getSelectedTab(state)];
export const getPanelLayoutGeneration: Selector<number> = (state) =>
  getApp(state).panelLayoutGeneration;
export const getLastVisibleThreadTabSlug: Selector<TabSlug> = (state) =>
  getApp(state).lastVisibleThreadTabSlug;
export const getTrackThreadHeights: Selector<{
  [key: ThreadsKey]: CssPixels,
}> = (state) => getApp(state).trackThreadHeights;
export const getIsNewlyPublished: Selector<boolean> = (state) =>
  getApp(state).isNewlyPublished;
export const getExperimental: Selector<ExperimentalFlags> = (state) =>
  getApp(state).experimental;
export const getBrowserConnectionStatus: Selector<BrowserConnectionStatus> = (
  state
) => getApp(state).browserConnectionStatus;
export const getBrowserConnection: Selector<BrowserConnection | null> =
  createSelector(getBrowserConnectionStatus, (status) =>
    status.status === 'ESTABLISHED' ? status.browserConnection : null
  );
export const getIsEventDelayTracksEnabled: Selector<boolean> = (state) =>
  getExperimental(state).eventDelayTracks;
export const getIsExperimentalCPUGraphsEnabled: Selector<boolean> = (state) =>
  getExperimental(state).cpuGraphs;
export const getIsExperimentalProcessCPUTracksEnabled: Selector<boolean> = (
  state
) => getExperimental(state).processCPUTracks;

export const getIsDragAndDropDragging: Selector<boolean> = (state) =>
  getApp(state).isDragAndDropDragging;
export const getIsDragAndDropOverlayRegistered: Selector<boolean> = (state) =>
  getApp(state).isDragAndDropOverlayRegistered;

export const getCurrentProfileUploadedInformation: Selector<
  UploadedProfileInformation | null
> = (state) => getApp(state).currentProfileUploadedInformation;

/**
 * Height of screenshot track is different depending on the view.
 */
export const getScreenshotTrackHeight: Selector<number> = createSelector(
  getTimelineTrackOrganization,
  (timelineTrackOrganization) => {
    switch (timelineTrackOrganization.type) {
      case 'active-tab':
        return ACTIVE_TAB_TRACK_SCREENSHOT_HEIGHT;
      case 'full':
      case 'origins':
        return FULL_TRACK_SCREENSHOT_HEIGHT;
      default:
        throw assertExhaustiveCheck(
          timelineTrackOrganization,
          `Unhandled TimelineTrackOrganization`
        );
    }
  }
);

/**
 * This selector takes all of the tracks, and deduces the height in CssPixels
 * of the timeline. This is here to calculate the max-height of the timeline
 * for the splitter component.
 *
 * The height of the component is determined by the sizing of each track in the list.
 * Most sizes are pretty static, and are set through values in the component. The only
 * tricky value to determine is the thread track. These values get reported to the store
 * and get added in here.
 */
export const getTimelineHeight: Selector<null | CssPixels> = createSelector(
  getTimelineTrackOrganization,
  getGlobalTracks,
  getLocalTracksByPid,
  getHiddenGlobalTracks,
  getHiddenLocalTracksByPid,
  getTrackThreadHeights,
  getActiveTabTimeline,
  getIsActiveTabResourcesPanelOpen,
  getScreenshotTrackHeight,
  (
    timelineTrackOrganization,
    globalTracks,
    localTracksByPid,
    hiddenGlobalTracks,
    hiddenLocalTracksByPid,
    trackThreadHeights,
    activeTabTimeline,
    isActiveTabResourcesPanelOpen,
    screenshotTrackHeight
  ) => {
    let height = TIMELINE_RULER_HEIGHT;
    const border = 1;
    switch (timelineTrackOrganization.type) {
      case 'origins': {
        return height + 500;
      }
      case 'active-tab': {
        if (activeTabTimeline.resources.length > 0) {
          // Active tab resources panel has a header and we should also add its
          // height if there is a panel there.
          height += ACTIVE_TAB_TIMELINE_RESOURCES_HEADER_HEIGHT;
        }

        // Add the height of the main track.
        // The thread tracks have enough complexity that it warrants measuring
        // them rather than statically using a value like the other tracks.
        const { threadsKey } = activeTabTimeline.mainTrack;
        const trackThreadHeight = trackThreadHeights[threadsKey];
        if (trackThreadHeight === undefined) {
          // The height isn't computed yet, return.
          return null;
        }
        height += trackThreadHeight + border;

        // Add the height of screenshot tracks.
        for (let i = 0; i < activeTabTimeline.screenshots.length; i++) {
          height += screenshotTrackHeight + border;
        }

        if (isActiveTabResourcesPanelOpen) {
          for (const resourceTrack of activeTabTimeline.resources) {
            switch (resourceTrack.type) {
              case 'sub-frame':
              case 'thread':
                {
                  // The thread tracks have enough complexity that it warrants measuring
                  // them rather than statically using a value like the other tracks.
                  const { threadIndex } = resourceTrack;
                  if (threadIndex === null) {
                    height += TRACK_PROCESS_BLANK_HEIGHT + border;
                  } else {
                    const trackThreadHeight = trackThreadHeights[threadIndex];
                    if (trackThreadHeight === undefined) {
                      // The height isn't computed yet, return.
                      return null;
                    }
                    height +=
                      trackThreadHeight +
                      ACTIVE_TAB_TIMELINE_RESOURCES_HEADER_HEIGHT +
                      border;
                  }
                }
                break;
              default:
                throw assertExhaustiveCheck(resourceTrack);
            }
          }
        }

        return height;
      }
      case 'full': {
        for (const [trackIndex, globalTrack] of globalTracks.entries()) {
          if (!hiddenGlobalTracks.has(trackIndex)) {
            switch (globalTrack.type) {
              case 'screenshots':
                height += screenshotTrackHeight + border;
                break;
              case 'visual-progress':
              case 'perceptual-visual-progress':
              case 'contentful-visual-progress':
                height += TRACK_VISUAL_PROGRESS_HEIGHT;
                break;
              case 'process': {
                // The thread tracks have enough complexity that it warrants measuring
                // them rather than statically using a value like the other tracks.
                const { mainThreadIndex } = globalTrack;
                if (mainThreadIndex === null) {
                  height += TRACK_PROCESS_BLANK_HEIGHT + border;
                } else {
                  const trackThreadHeight = trackThreadHeights[mainThreadIndex];
                  if (trackThreadHeight === undefined) {
                    // The height isn't computed yet, return.
                    return null;
                  }
                  height += trackThreadHeight + border;
                }
                break;
              }
              default:
                throw assertExhaustiveCheck(globalTrack);
            }
          }
        }

        // Figure out which PIDs are hidden.
        const hiddenPids = new Set();
        for (const trackIndex of hiddenGlobalTracks) {
          const globalTrack = globalTracks[trackIndex];
          if (globalTrack.type === 'process') {
            hiddenPids.add(globalTrack.pid);
          }
        }

        for (const [pid, localTracks] of localTracksByPid) {
          if (hiddenPids.has(pid)) {
            // This track is hidden already.
            continue;
          }
          for (const [trackIndex, localTrack] of localTracks.entries()) {
            const hiddenLocalTracks = ensureExists(
              hiddenLocalTracksByPid.get(pid),
              'Could not look up the hidden local tracks from the given PID'
            );
            if (!hiddenLocalTracks.has(trackIndex)) {
              switch (localTrack.type) {
                case 'thread':
                  {
                    // The thread tracks have enough complexity that it warrants measuring
                    // them rather than statically using a value like the other tracks.
                    const trackThreadHeight =
                      trackThreadHeights[localTrack.threadIndex];
                    if (trackThreadHeight === undefined) {
                      // The height isn't computed yet, return.
                      return null;
                    }
                    height += trackThreadHeight + border;
                  }

                  break;
                case 'network':
                  height += TRACK_NETWORK_HEIGHT + border;
                  break;
                case 'memory':
                  height += TRACK_MEMORY_HEIGHT + border;
                  break;
                case 'event-delay':
                  height += TRACK_EVENT_DELAY_HEIGHT + border;
                  break;
                case 'ipc':
                  height += TRACK_IPC_HEIGHT + border;
                  break;
                case 'process-cpu':
                case 'power':
                  height += TRACK_PROCESS_CPU_HEIGHT + border;
                  break;
                default:
                  throw assertExhaustiveCheck(localTrack);
              }
            }
          }
        }
        return height;
      }
      default:
        throw assertExhaustiveCheck(
          timelineTrackOrganization,
          `Unhandled TimelineTrackOrganization`
        );
    }
  }
);

/**
 * This selector lets us know if it is safe to load a new profile. If
 * the app is already busy loading a profile, this selector returns
 * false.
 *
 * Used by the drag and drop component in order to determine if it can
 * load a dropped profile file.
 */
export const getIsNewProfileLoadAllowed: Selector<boolean> = createSelector(
  getView,
  getDataSource,
  getZipFileState,
  (view, dataSource, zipFileState) => {
    const appPhase = view.phase;
    const zipPhase = zipFileState.phase;
    const isLoading =
      (appPhase === 'INITIALIZING' && dataSource !== 'none') ||
      zipPhase === 'PROCESS_PROFILE_FROM_ZIP_FILE';
    return !isLoading;
  }
);

/**
 * Height of screenshot track is different depending on the view.
 */
export const getTimelineMarginLeft: Selector<number> = createSelector(
  getTimelineTrackOrganization,
  (timelineTrackOrganization) => {
    switch (timelineTrackOrganization.type) {
      case 'active-tab':
        return ACTIVE_TAB_TIMELINE_MARGIN_LEFT;
      case 'full':
      case 'origins':
        return TIMELINE_MARGIN_LEFT;
      default:
        throw assertExhaustiveCheck(
          timelineTrackOrganization,
          `Unhandled TimelineTrackOrganization`
        );
    }
  }
);

/**
 * Returns the indexes of categories that are opened in the sidebar,
 * for every category
 */
export const getSidebarOpenCategories: Selector<Map<string, Set<number>>> =
  createSelector(getApp, (app) => app.sidebarOpenCategories);
