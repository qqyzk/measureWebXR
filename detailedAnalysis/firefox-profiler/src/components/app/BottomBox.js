/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import React from 'react';
import classNames from 'classnames';

import { SourceView } from '../shared/SourceView';
import {
  getSourceViewFile,
  getSourceViewActivationGeneration,
} from 'firefox-profiler/selectors/url-state';
import {
  selectedThreadSelectors,
  selectedNodeSelectors,
} from 'firefox-profiler/selectors/per-thread';
import { closeBottomBox } from 'firefox-profiler/actions/profile-view';
import { parseFileNameFromSymbolication } from 'firefox-profiler/utils/special-paths';
import { getSourceViewSource } from 'firefox-profiler/selectors/sources';
import { getPreviewSelection } from 'firefox-profiler/selectors/profile';
import { assertExhaustiveCheck } from 'firefox-profiler/utils/flow';
import explicitConnect from 'firefox-profiler/utils/connect';

import type { ConnectedProps } from 'firefox-profiler/utils/connect';
import type { LineTimings, FileSourceStatus } from 'firefox-profiler/types';

import { Localized } from '@fluent/react';

import './BottomBox.css';

type StateProps = {|
  +sourceViewFile: string | null,
  +sourceViewSource: FileSourceStatus | void,
  +globalLineTimings: LineTimings,
  +selectedCallNodeLineTimings: LineTimings,
  +sourceViewActivationGeneration: number,
  +disableOverscan: boolean,
|};

type DispatchProps = {|
  +closeBottomBox: typeof closeBottomBox,
|};

type Props = ConnectedProps<{||}, StateProps, DispatchProps>;

type SourceStatusOverlayProps = {| status: FileSourceStatus |};

function SourceStatusOverlay({ status }: SourceStatusOverlayProps) {
  switch (status.type) {
    case 'AVAILABLE':
      return null; // No overlay if we have source code.
    case 'LOADING': {
      const { source } = status;
      switch (source.type) {
        case 'URL': {
          const { url } = source;
          let host;
          try {
            host = new URL(url).host;
          } catch (e) {
            host = url;
          }
          return (
            <Localized id="SourceView--loading-url" vars={{ host }}>
              <div className="sourceStatusOverlay loading">
                {`Waiting for ${host}…`}
              </div>
            </Localized>
          );
        }
        case 'BROWSER_CONNECTION': {
          return (
            <Localized id="SourceView--loading-browser-connection">
              <div className="sourceStatusOverlay loading">
                Waiting for browser…
              </div>
            </Localized>
          );
        }
        default:
          throw assertExhaustiveCheck(source.type);
      }
    }
    case 'ERROR': {
      return (
        <div className="sourceStatusOverlay error">
          <div>
            <Localized id="SourceView--source-not-available-title">
              <h3>Source not available</h3>
            </Localized>
            <Localized
              id="SourceView--source-not-available-text"
              elems={{
                a: (
                  <a
                    href="https://github.com/firefox-devtools/profiler/issues/3741"
                    target="_blank"
                    rel="noopener noreferrer"
                  />
                ),
              }}
            >
              <p>
                See
                <a
                  href="https://github.com/firefox-devtools/profiler/issues/3741"
                  target="_blank"
                  rel="noopener noreferrer"
                >
                  issue #3741
                </a>
                for supported scenarios and planned improvements.
              </p>
            </Localized>
            <ul>
              {status.errors.map((error, key) => {
                switch (error.type) {
                  case 'NO_KNOWN_CORS_URL': {
                    return (
                      <Localized key={key} id="SourceView--no-known-cors-url">
                        <li>No known cross-origin-accessible URL.</li>
                      </Localized>
                    );
                  }
                  case 'NETWORK_ERROR': {
                    const { url, networkErrorMessage } = error;
                    return (
                      <Localized
                        key={key}
                        id="SourceView--network-error-when-obtaining-source"
                        vars={{ url, networkErrorMessage }}
                      >
                        <li>{`There was a network error when fetching the URL ${url}: ${networkErrorMessage}`}</li>
                      </Localized>
                    );
                  }
                  case 'BROWSER_CONNECTION_ERROR': {
                    const { browserConnectionErrorMessage } = error;
                    return (
                      <Localized
                        key={key}
                        id="SourceView--browser-connection-error-when-obtaining-source"
                        vars={{ browserConnectionErrorMessage }}
                      >
                        <li>{`Could not query the browser’s symbolication API: ${browserConnectionErrorMessage}`}</li>
                      </Localized>
                    );
                  }
                  case 'BROWSER_API_ERROR': {
                    const { apiErrorMessage } = error;
                    return (
                      <Localized
                        id="SourceView--browser-api-error-when-obtaining-source"
                        vars={{ apiErrorMessage }}
                      >
                        <li>{`The browser’s symbolication API returned an error: ${apiErrorMessage}`}</li>
                      </Localized>
                    );
                  }
                  case 'SYMBOL_SERVER_API_ERROR': {
                    const { apiErrorMessage } = error;
                    return (
                      <Localized
                        id="SourceView--local-symbol-server-api-error-when-obtaining-source"
                        vars={{ apiErrorMessage }}
                      >
                        <li>{`The local symbol server’s symbolication API returned an error: ${apiErrorMessage}`}</li>
                      </Localized>
                    );
                  }
                  case 'NOT_PRESENT_IN_ARCHIVE': {
                    const { url, pathInArchive } = error;
                    return (
                      <Localized
                        id="SourceView--not-in-archive-error-when-obtaining-source"
                        vars={{ url, pathInArchive }}
                      >
                        <li>{`The file ${pathInArchive} was not found in the archive from ${url}.`}</li>
                      </Localized>
                    );
                  }
                  case 'ARCHIVE_PARSING_ERROR': {
                    const { url, parsingErrorMessage } = error;
                    return (
                      <Localized
                        id="SourceView--archive-parsing-error-when-obtaining-source"
                        vars={{ url, parsingErrorMessage }}
                      >
                        <li>{`The archive at ${url} could not be parsed: ${parsingErrorMessage}`}</li>
                      </Localized>
                    );
                  }
                  default:
                    throw assertExhaustiveCheck(error.type);
                }
              })}
            </ul>
          </div>
        </div>
      );
    }
    default:
      throw assertExhaustiveCheck(status.type);
  }
}

class BottomBoxImpl extends React.PureComponent<Props> {
  _sourceView = React.createRef<SourceView>();

  _onClickCloseButton = () => {
    this.props.closeBottomBox();
  };

  render() {
    const {
      sourceViewFile,
      sourceViewSource,
      globalLineTimings,
      disableOverscan,
      sourceViewActivationGeneration,
      selectedCallNodeLineTimings,
    } = this.props;
    const source =
      sourceViewSource && sourceViewSource.type === 'AVAILABLE'
        ? sourceViewSource.source
        : '';
    const path =
      sourceViewFile !== null
        ? parseFileNameFromSymbolication(sourceViewFile).path
        : null;
    return (
      <div className="bottom-box">
        <div className="bottom-box-bar">
          <h3 className="bottom-box-title">{path ?? '(no file selected)'}</h3>
          <Localized id="SourceView--close-button" attrs={{ title: true }}>
            <button
              className={classNames(
                'bottom-close-button',
                'photon-button',
                'photon-button-ghost'
              )}
              title="Close the source view"
              type="button"
              onClick={this._onClickCloseButton}
            />
          </Localized>
        </div>
        <div className="bottom-main" id="bottom-main">
          {sourceViewFile !== null ? (
            <SourceView
              disableOverscan={disableOverscan}
              timings={globalLineTimings}
              source={source}
              filePath={path}
              scrollToHotSpotGeneration={sourceViewActivationGeneration}
              hotSpotTimings={selectedCallNodeLineTimings}
              ref={this._sourceView}
            />
          ) : null}
          {sourceViewSource !== undefined &&
          sourceViewSource.type !== 'AVAILABLE' ? (
            <SourceStatusOverlay status={sourceViewSource} />
          ) : null}
        </div>
      </div>
    );
  }
}

export const BottomBox = explicitConnect<{||}, StateProps, DispatchProps>({
  mapStateToProps: (state) => ({
    sourceViewFile: getSourceViewFile(state),
    sourceViewSource: getSourceViewSource(state),
    globalLineTimings: selectedThreadSelectors.getSourceViewLineTimings(state),
    selectedCallNodeLineTimings:
      selectedNodeSelectors.getSourceViewLineTimings(state),
    sourceViewActivationGeneration: getSourceViewActivationGeneration(state),
    disableOverscan: getPreviewSelection(state).isModifying,
  }),
  mapDispatchToProps: {
    closeBottomBox,
  },
  component: BottomBoxImpl,
});
