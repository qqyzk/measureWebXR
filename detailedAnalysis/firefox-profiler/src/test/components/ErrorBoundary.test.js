/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as React from 'react';
import { screen } from '@testing-library/react';
import { stripIndent } from 'common-tags';

import { render } from 'firefox-profiler/test/fixtures/testing-library';
import { ErrorBoundary } from '../../components/app/ErrorBoundary';
import { withAnalyticsMock } from '../fixtures/mocks/analytics';
import { fireFullClick } from '../fixtures/utils';

describe('app/ErrorBoundary', function () {
  const childComponentText = 'This is a child component';
  const friendlyErrorMessage = 'Oops, there was an error';
  const technicalErrorMessage = 'This is an error.';
  const ThrowingComponent = () => {
    throw new Error(technicalErrorMessage);
  };

  function setupComponent(childComponent) {
    const spy = jest.spyOn(console, 'error').mockImplementation(() => {});
    const results = render(
      <ErrorBoundary message={friendlyErrorMessage}>
        {childComponent}
      </ErrorBoundary>
    );
    return { spy, ...results };
  }

  it('matches the snapshot', () => {
    const { container } = setupComponent(<ThrowingComponent />);

    // We need to change the textContent of the stack, so that path information
    // isn't tied to a specific environment.
    const stack = screen.getByText(/at ThrowingComponent/);
    stack.textContent = stack.textContent
      .replace(/\\/g, '/') // normalizes Windows paths
      .replace(/\(.*\/src/g, '(REDACTED)/src') // Removes the home directory
      // Removes the home directory and line / column numbers of vendored files.
      .replace(
        /\(.*\/node_modules\/([^:]*):.+/g,
        '(REDACTED)/node_modules/$1:(REDACTED)'
      );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('shows the component children when there is no error', () => {
    const { getByText } = setupComponent(childComponentText);
    expect(getByText(childComponentText)).toBeInTheDocument();
  });

  it('shows the error message children when the component throws error', () => {
    const { getByText } = setupComponent(<ThrowingComponent />);
    expect(getByText(friendlyErrorMessage)).toBeInTheDocument();
  });

  it('surfaces the error via console.error', () => {
    setupComponent(<ThrowingComponent />);
    expect(console.error).toHaveBeenCalled();
  });

  it('can click the component to view the full details', () => {
    const { getByText, getByTestId } = setupComponent(<ThrowingComponent />);

    // The technical error isn't visible yet.
    expect(getByTestId('error-technical-details')).toHaveClass('hide');

    // Click the button to expand the details.
    fireFullClick(getByText('View full error details'));

    // The technical error now exists.
    expect(getByTestId('error-technical-details')).not.toHaveClass('hide');
  });

  it('reports errors to the analytics', () => {
    withAnalyticsMock(() => {
      setupComponent(<ThrowingComponent />);
      expect(self.ga).toHaveBeenCalledWith('send', 'exception', {
        exDescription: expect.stringMatching(
          // This regexp looks a  bit scary, but all it does is avoiding
          // matching on things that depends on the environment (like path
          // names) and path separators (unixes use `/` but windows uses `\`).
          new RegExp(stripIndent`
              Error: This is an error\\.

                  at ThrowingComponent \\(.*[/\\\\]ErrorBoundary.test.js:20:11\\)
                  at ErrorBoundary \\(.*[/\\\\]ErrorBoundary.js:28:66\\)
                  at LocalizationProvider \\(.*[/\\\\]@fluent[/\\\\]react[/\\\\]index.js:.*\\)
          `)
        ),
        exFatal: true,
      });
    });
  });
});
