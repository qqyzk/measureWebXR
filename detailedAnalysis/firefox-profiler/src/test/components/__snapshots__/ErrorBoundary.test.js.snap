// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/ErrorBoundary matches the snapshot 1`] = `
<div
  class="appErrorBoundary"
>
  <div
    class="appErrorBoundaryContents"
  >
    <div
      class="photon-message-bar photon-message-bar-error photon-message-bar-inner-content appErrorBoundaryMessage"
    >
      <div
        class="photon-message-bar-inner-text"
      >
        Oops, there was an error
      </div>
      <button
        aria-expanded="false"
        class="photon-button photon-button-micro photon-message-bar-action-button"
        type="button"
      >
        View full error details
      </button>
    </div>
    <div
      class="appErrorBoundaryDetails hide"
      data-testid="error-technical-details"
    >
      <div>
        Error: This is an error.
      </div>
      <div>
        
    at ThrowingComponent (REDACTED)/src/test/components/ErrorBoundary.test.js:20:11)
    at ErrorBoundary (REDACTED)/src/components/app/ErrorBoundary.js:28:66)
    at LocalizationProvider (REDACTED)/node_modules/@fluent/react/index.js:(REDACTED)
      </div>
    </div>
  </div>
</div>
`;
